include/graphblas/reference/blas3.hpp

```diff
@@ -916,6 +916,206 @@ namespace grb {
 		return ret;
 	}
 
+
+	/**
+	 * A masked outer product of two vectors. Assuming vectors \a u and \a v are oriented
+	 * column-wise, the result matrix \a A will contain \f$ uv^T \f$, masked to non-zero values from \a mask.
+	 */
+	template<
+		Descriptor descr = descriptors::no_operation,
+		class Operator,
+		typename InputType1, typename InputType2, typename OutputType,
+		typename Coords,
+		typename RIT, typename CIT, typename NIT
+	>
+	RC maskedOuter(
+		Matrix< OutputType, reference, RIT, CIT, NIT > &A,
+		const Matrix< OutputType, reference, RIT, CIT, NIT > &mask,
+		const Vector< InputType1, reference, Coords > &u,
+		const Vector< InputType2, reference, Coords > &v,
+		const Operator &mul = Operator(),
+		const Phase &phase = EXECUTE,
+		const typename std::enable_if<
+			grb::is_operator< Operator >::value &&
+			!grb::is_object< InputType1 >::value &&
+			!grb::is_object< InputType2 >::value &&
+			!grb::is_object< OutputType >::value,
+			void >::type * const = nullptr
+	) {
+		// static checks
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D1, InputType1 >::value
+			), "grb::maskedOuter",
+			"called with a prefactor vector that does not match the first domain "
+			"of the given multiplication operator" );
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D2, InputType2 >::value
+			), "grb::maskedOuter",
+			"called with a postfactor vector that does not match the first domain "
+			"of the given multiplication operator" );
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D3, OutputType >::value
+			), "grb::maskedOuter",
+			"called with an output matrix that does not match the output domain of "
+			"the given multiplication operator" );
+		static_assert( ( !(descr & descriptors::invert_mask)
+			),
+			"grb::maskedOuter: invert_mask descriptor cannot be used ");
+#ifdef _DEBUG
+		std::cout << "In grb::maskedOuter (reference)\n";
+#endif
+
+		const size_t nrows = size( u );
+		const size_t ncols = size( v );
+
+		const size_t m = grb::nrows( mask );
+		const size_t n = grb::ncols( mask );
+
+		assert( phase != TRY );
+		if( nrows != grb::nrows( A ) || nrows != m ) {
+			return MISMATCH;
+		}
+
+		if( ncols != grb::ncols( A ) || ncols != n ) {
+			return MISMATCH;
+		}
+
+
+		const auto &mask_raw = internal::getCRS( mask );
+
+		size_t nzc = 0;
+		for( size_t i = 0; i < nrows; ++i ) {
+			if( internal::getCoordinates( u ).assigned( i ) ) {
+				for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
+					const size_t k_col = mask_raw.row_index[ k ];
+					if( internal::getCoordinates( v ).assigned( k_col ) ) {
+						nzc++;
+					}
+				}
+			}
+		}
```

Comment:
If the matrix contains many sparse rows, iterating over the assigned coordinates of `u` would be more efficient. 

```suggestion
		for( size_t i : internal::getCoordinates( u ) ) {
			for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
				const size_t k_col = mask_raw.row_index[ k ];
				if( internal::getCoordinates( v ).assigned( k_col ) ) {
					nzc++;
				}
			}
		}
```

On the other hand, you can no longer use openMP on the outer foreach-loop. This leads to the second suggestion:

```suggestion
#ifdef _H_GRB_REFERENCE_OMP_BLAS3
		#pragma omp parallel for reduction(+:nzc)
#endif
		for( size_t i = 0; i < nrows; ++i ) {
			if( internal::getCoordinates( u ).assigned( i ) ) {
				for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
					const size_t k_col = mask_raw.row_index[ k ];
					if( internal::getCoordinates( v ).assigned( k_col ) ) {
						nzc++;
					}
				}
			}
		}
```

Reply:
Thanks, I opted for the second suggestion here